src/main/scala/com/redislabs/provider/redis/sql/DefaultSource.scala

```diff
@@ -0,0 +1,180 @@
+package com.redislabs.provider.redis.sql
+
+import scala.collection.JavaConversions._
+import com.redislabs.provider.redis._
+import com.redislabs.provider.redis.rdd.{Keys, RedisKeysRDD}
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.{DataFrame, SQLContext}
+import org.apache.spark.sql.catalyst.expressions.Row
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import redis.clients.jedis.Protocol
+import redis.clients.util.JedisClusterCRC16
+import java.security.MessageDigest
+
+
+case class RedisRelation(parameters: Map[String, String], userSchema: StructType)
+                        (@transient val sqlContext: SQLContext)
+  extends BaseRelation with PrunedFilteredScan with InsertableRelation with Keys {
+
+  val tableName: String = parameters.getOrElse("table", "PANIC")
+
+  val redisConfig: RedisConfig = {
+    new RedisConfig({
+        if ((parameters.keySet & Set("host", "port", "auth", "dbNum", "timeout")).size == 0) {
+          new RedisEndpoint(sqlContext.sparkContext.getConf)
+        } else {
+          val host = parameters.getOrElse("host", Protocol.DEFAULT_HOST)
+          val port = parameters.getOrElse("port", Protocol.DEFAULT_PORT.toString).toInt
+          val auth = parameters.getOrElse("auth", null)
+          val dbNum = parameters.getOrElse("dbNum", Protocol.DEFAULT_DATABASE.toString).toInt
+          val timeout = parameters.getOrElse("timeout", Protocol.DEFAULT_TIMEOUT.toString).toInt
+          new RedisEndpoint(host, port, auth, dbNum, timeout)
+        }
+      }
+    )
+  }
+
+  val partitionNum: Int = parameters.getOrElse("partitionNum", 3.toString).toInt
+
+  val schema = userSchema
+
+  def getNode(key: String): RedisNode = {
+    val slot = JedisClusterCRC16.getSlot(key)
+    /* Master only */
+    redisConfig.hosts.filter(node => { node.startSlot <= slot && node.endSlot >= slot }).filter(_.idx == 0)(0)
+  }
+  def insert(data: DataFrame, overwrite: Boolean): Unit = {
+    data.foreach{
+      row => {
+        val key = tableName + ":" + MessageDigest.getInstance("MD5").digest(System.currentTimeMillis.toString.getBytes)
+        val conn = getNode(key).endpoint.connect
+        conn.hmset(key, row.getValuesMap(row.schema.fieldNames).map(x => (x._1, x._2.toString)))
+        conn.close
+      }
+    }
+  }
+
+  def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
+    new RedisKeysRDD(sqlContext.sparkContext, redisConfig, tableName + ":*", partitionNum, null).
+      mapPartitions {
+        partition: Iterator[String] => {
+          groupKeysByNode(redisConfig.hosts, partition).flatMap {
+            x => {
+              val conn = x._1.endpoint.connect()
+              val rowKeys: Array[String] = filterKeysByType(conn, x._2, "hash")
+              val res = rowKeys.map {
+                key => {
+                  val res = conn.hmget(key, schema.fieldNames: _*)
```

Comment:
But if the filters depends on other fields we didn't require, we should also know them